rx_signal_processing/decimate.cu

```diff
@@ -1,221 +1,693 @@
-
-    auto filter_tap_num = threadIdx.x;
-    auto num_filter_taps = blockDim.x;
-    cuComplex total_sum = data[tap_offset];
 
+#include <cuComplex.h> //cuComplex type and all cuCmulf/cuCaddf functions.
+#include <iostream>
+#include <stdint.h>
+#include "decimate.hpp"
+//This keeps the contained functions local to this file.
+namespace {
+  /**
+   * @brief      Creates a new set of grid dimensions for a bandpass decimate CUDA kernel.
+   *
+   * @param[in]  num_samples   Number of input samples.
+   * @param[in]  dm_rate       Decimation rate.
+   * @param[in]  num_antennas  Number of antennas for which there are samples.
+   *
+   * @return     New grid dimensions for the kernel.
+   */
+  dim3 create_bandpass_grid(uint32_t num_samples, uint32_t dm_rate, uint32_t num_antennas)
+  {
+    auto num_blocks_x = num_samples/dm_rate;
+    auto num_blocks_y = num_antennas;
+    auto num_blocks_z = 1;
+    std::cout << "    Grid size: " << num_blocks_x << " x " << num_blocks_y << " x "
+      << num_blocks_z << std::endl;
+    dim3 dimGrid(num_blocks_x,num_blocks_y,num_blocks_z);
 
-    if ((num_filter_taps >= 512) && (filter_tap_num < 256))
-    {
-        total_sum = cuCaddf(total_sum,data[tap_offset  + 256]);
-        data[tap_offset] = total_sum; // REVIEW #25 Is it necessary for speed to have two '=' statements on one line? it took a while to see the second one, therefore more confusing. split into two lines
-    }                                       // REPLY okay
-
-    __syncthreads();
+    return dimGrid;
+  }
+
+  /**
+   * @brief      Creates a new set of block dimensions for a bandpass decimate CUDA kernel.
+   *
+   * @param[in]  num_taps_per_filter  Number of taps per filter.
+   * @param[in]  num_freqs            Number of receive frequencies.
+   *
+   * @return     New block dimensions for the kernel.
+   */
+  dim3 create_bandpass_block(uint32_t num_taps_per_filter, uint32_t num_freqs)
+  {
+    auto num_threads_x = num_taps_per_filter;
+    auto num_threads_y = num_freqs;
+    auto num_threads_z = 1;
+    std::cout << "    Block size: " << num_threads_x << " x " << num_threads_y << " x "
+      << num_threads_z << std::endl;
+    dim3 dimBlock(num_threads_x,num_threads_y,num_threads_z);
 
-    if ((num_filter_taps >= 256) && (filter_tap_num < 128))
-    {
-        total_sum = cuCaddf(total_sum, data[tap_offset + 128]);
-        data[tap_offset] = total_sum;
-    }
+    return dimBlock;
+  }
+
+  /**
+   * @brief      Creates a new set of grid dimensions for a lowpass decimate CUDA kernel.
+   *
+   * @param[in]  num_samples   Number of input samples in a frequency dataset.
+   * @param[in]  dm_rate       Decimation rate.
+   * @param[in]  num_antennas  Number of antennas for which there are samples.
+   * @param[in]  num_freqs     Number of receive frequencies.
+   *
+   * @return     New grid dimensions for the kernel.
+   */
+  dim3 create_lowpass_grid(uint32_t num_samples, uint32_t dm_rate, uint32_t num_antennas,
```

Comment:
I'm not sure if we will ever have filters that large, but its there if it ever will be used. Regardless of filter size, you will take a performance hit to use the same block structure for bandpass. You would need a seperate kernel instance for each RX frequency, with many of the threads going unused. The bandpass structure, either 1024 or 2048, is going to allow you to run all your bandpass filters in parallel with the 2 dimensional block structure. The only difference between 1024 and 2048 versions are 2 taps are performed by 1 thread in the 2048 kernels. That is the same in both lowpass and bandpass kernels.